### PR TITLE
Do not allow --extra-jwt-issuers and --allowed-group together.

### PR DESCRIPTION
**Pull request by theGOTOguy**
As per the discussion in #1243, oauth2-proxy will not support this behavior until v8.

## Description

Adds a check to see if both --extra-jwt-issuers and --allowed-group were provided, and fails with a clear error message if they were.

## Motivation and Context

It is currently impossible that any --extra-jwt-issuers would be respected if any --allowed-group is provided.  Instead of leaving this behavior undocumented and non-functional, disallow these flags to be provided together to save users some time.

## How Has This Been Tested?

Built a Docker image with the change and tested the behavior manually using `--extra-jwt-issuers` and `--allowed-group`

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Review by NickMeves — COMMENTED** [review]
I'm inclined not to merge this 🤔 

I'm a little out of practice with the nitty gritty -- but I think this will break this architecture I used at Greenhouse to allow machine + human users concurrently:

https://medium.com/in-the-weeds/service-to-service-authentication-on-kubernetes-94dcb8216cdc

The allowed groups only applies to the human interactive flow with the main provider, right? (and we can then give machine OIDC issuers as extra JWTs since that codepath uses a completely different authorization route that doesn't check allowed groups).

**Comment by theGOTOguy** [comment]
If you provided both `--extra-jwt-issuers` and `--allowed-group`, the `--extra-jwt-issuers` are expected to match a group in `--allowed-group`.  But, since the groups for the `--extra-jwt-issuers` are never loaded, then they can never possibly match any of the `--allowed-group`s and as such `--extra-jwt-issuers` are effectively ignored.

Enabling both human and machine users concurrently is exactly my use case, and is in fact the reason I had to fork OAuth2 proxy:  Humans come in through the main issuer and their group is correctly loaded.  Scripts running through the CLI come through the `--extra-jwt-issuers`.  However, even though these service accounts are members of the `--allowed-group`, they are rejected because said groups are never loaded for the `--extra-jwt-issuers`.

I think we agreed in the discussion for #1243 that there was no way that any of the `--extra-jwt-issuers` could possibly be respected if `--allowed-group` is provided, and the rationale for this PR is simply to avoid user confusion should anyone try to provide both together.

**Comment by NickMeves** [comment]
Thanks for clarifying! Sorry, its been so long I've forgotten 😅 

I see your point 👍  I traced the code, even though extra JWTs has its own Session generator codepath, every request calls `getAuthenticedSession` which calls the provider `Authorize` method.

Looking back at the underlying issue, I stick by my opinion to wait until we deprecate this structure completely in v8 and roll out the replacement: https://github.com/oauth2-proxy/oauth2-proxy/issues/1243#issuecomment-864346754

@JoelSpeed and I have also discussed moving away from provider specific `Authorize` implememtations and moving toward a central rules based engine that is provider agnostic and solely works on sessions.

File: main.go [hunk 1]
```diff
@@ -50,6 +50,14 @@ func main() {
 		return
 	}
 
+        if opts.ExtraJwtIssuers != nil {
```
Comment (NickMeves) [inline main.go:53]: If you really want to implement this in the interim until deprecation, this isn't the appropriate spot.

This belongs in the validation package and likely needs to be flag & alpha options aware? (Hence my recommendation to punt until we completely deprecate this structure).

Comment (theGOTOguy) [reply main.go]: Moved it to validation options.

**Comment by theGOTOguy** [comment]
I absolutely respect your rationale for waiting for v8 to fix the issue.   I would just make sure that the existing behavior, which may be surprising or confusing, is documented.  My preference in that regard is clearly to outright refuse to allow a flag combination we know isn't going to work as the user expects.  In my case, before I figured this out I had to re-read the docs, check my configuration, capture and confirm that my JWTs are as expected, then printf-debug the code before I finally discovered that the `--extra-jwt-issuers` could never work when the `--allowed-group` flag is specified.  

To test, run `docker build .` 
Note the image hash and use it below.  In my case, it was `de09d99144e6`.

No `--allowed-groups` or `--extra-jwt-issuers`:
```bash
docker run -p 4180:4180 \
           -e OAUTH2_PROXY_COOKIE_SECRET=secretsecretsecr \
           -e OAUTH2_PROXY_CLIENT_ID=clientid \
           -e OAUTH2_PROXY_CLIENT_SECRET=secret \
           -e OAUTH2_PROXY_EMAIL_DOMAINS=* \
           -e OAUTH2_PROXY_UPSTREAMS=http://localhost:8080 \
           -e OAUTH2_PROXY_LOGIN_URL=https://oauth.mocklab.io/oauth/authorize \
           -e OAUTH2_PROXY_OIDC_JWKS_URL=https://oauth.mocklab.io/.well-known/jwks.json \
           -e OAUTH2_PROXY_REDEEM_URL=https://oauth.mocklab.io/oauth/token \
           -e OAUTH2_PROXY_COOKIE_SECURE=false \
           -e OAUTH2_PROXY_PROFILE_URL=https://oauth.mocklab.io/userinfo \
           -e OAUTH2_PROXY_HTTP_ADDRESS=0.0.0.0:4180 \
           -e OAUTH2_PROXY_PROVIDER=oidc \
           -e OAUTH2_PROXY_OIDC_ISSUER_URL=https://oauth.mocklab.io \
           -e OAUTH2_PROXY_SKIP_OIDC_DISCOVERY=true \
           -e OAUTH2_PROXY_SKIP_JWT_BEARER_TOKENS=true \
           -e OAUTH2_PROXY_REDIRECT_URL=http://localhost:4180/oauth2/callback \
           --network="host" \
           de09d99144e6
```

Just `--allowed-group`
```bash
docker run -p 4180:4180 \
           -e OAUTH2_PROXY_COOKIE_SECRET=secretsecretsecr \
           -e OAUTH2_PROXY_CLIENT_ID=clientid \
           -e OAUTH2_PROXY_CLIENT_SECRET=secret \
           -e OAUTH2_PROXY_EMAIL_DOMAINS=* \
           -e OAUTH2_PROXY_UPSTREAMS=http://localhost:8080 \
           -e OAUTH2_PROXY_LOGIN_URL=https://oauth.mocklab.io/oauth/authorize \
           -e OAUTH2_PROXY_OIDC_JWKS_URL=https://oauth.mocklab.io/.well-known/jwks.json \
           -e OAUTH2_PROXY_REDEEM_URL=https://oauth.mocklab.io/oauth/token \
           -e OAUTH2_PROXY_COOKIE_SECURE=false \
           -e OAUTH2_PROXY_PROFILE_URL=https://oauth.mocklab.io/userinfo \
           -e OAUTH2_PROXY_HTTP_ADDRESS=0.0.0.0:4180 \
           -e OAUTH2_PROXY_PROVIDER=oidc \
           -e OAUTH2_PROXY_OIDC_ISSUER_URL=https://oauth.mocklab.io \
           -e OAUTH2_PROXY_SKIP_OIDC_DISCOVERY=true \
           -e OAUTH2_PROXY_SKIP_JWT_BEARER_TOKENS=true \
           -e OAUTH2_PROXY_REDIRECT_URL=http://localhost:4180/oauth2/callback \
           -e OAUTH2_PROXY_ALLOWED_GROUPS=fake \
           --network="host" \
           de09d99144e6
```

Just `--extra-jwt-issuers`:
```bash
docker run -p 4180:4180 \
           -e OAUTH2_PROXY_COOKIE_SECRET=secretsecretsecr \
           -e OAUTH2_PROXY_CLIENT_ID=clientid \
           -e OAUTH2_PROXY_CLIENT_SECRET=secret \
           -e OAUTH2_PROXY_EMAIL_DOMAINS=* \
           -e OAUTH2_PROXY_UPSTREAMS=http://localhost:8080 \
           -e OAUTH2_PROXY_LOGIN_URL=https://oauth.mocklab.io/oauth/authorize \
           -e OAUTH2_PROXY_OIDC_JWKS_URL=https://oauth.mocklab.io/.well-known/jwks.json \
           -e OAUTH2_PROXY_REDEEM_URL=https://oauth.mocklab.io/oauth/token \
           -e OAUTH2_PROXY_COOKIE_SECURE=false \
           -e OAUTH2_PROXY_PROFILE_URL=https://oauth.mocklab.io/userinfo \
           -e OAUTH2_PROXY_HTTP_ADDRESS=0.0.0.0:4180 \
           -e OAUTH2_PROXY_PROVIDER=oidc \
           -e OAUTH2_PROXY_OIDC_ISSUER_URL=https://oauth.mocklab.io \
           -e OAUTH2_PROXY_SKIP_OIDC_DISCOVERY=true \
           -e OAUTH2_PROXY_SKIP_JWT_BEARER_TOKENS=true \
           -e OAUTH2_PROXY_REDIRECT_URL=http://localhost:4180/oauth2/callback \
           -e OAUTH2_PROXY_EXTRA_JWT_ISSUERS=https://oauth.mocklab.io/=https://oauth.mocklab.io/ \
           --network="host" \
           de09d99144e6
```

Both Together:
```bash
docker run -p 4180:4180 \
           -e OAUTH2_PROXY_COOKIE_SECRET=secretsecretsecr \
           -e OAUTH2_PROXY_CLIENT_ID=clientid \
           -e OAUTH2_PROXY_CLIENT_SECRET=secret \
           -e OAUTH2_PROXY_EMAIL_DOMAINS=* \
           -e OAUTH2_PROXY_UPSTREAMS=http://localhost:8080 \
           -e OAUTH2_PROXY_LOGIN_URL=https://oauth.mocklab.io/oauth/authorize \
           -e OAUTH2_PROXY_OIDC_JWKS_URL=https://oauth.mocklab.io/.well-known/jwks.json \
           -e OAUTH2_PROXY_REDEEM_URL=https://oauth.mocklab.io/oauth/token \
           -e OAUTH2_PROXY_COOKIE_SECURE=false \
           -e OAUTH2_PROXY_PROFILE_URL=https://oauth.mocklab.io/userinfo \
           -e OAUTH2_PROXY_HTTP_ADDRESS=0.0.0.0:4180 \
           -e OAUTH2_PROXY_PROVIDER=oidc \
           -e OAUTH2_PROXY_OIDC_ISSUER_URL=https://oauth.mocklab.io \
           -e OAUTH2_PROXY_SKIP_OIDC_DISCOVERY=true \
           -e OAUTH2_PROXY_SKIP_JWT_BEARER_TOKENS=true \
           -e OAUTH2_PROXY_REDIRECT_URL=http://localhost:4180/oauth2/callback \
           -e OAUTH2_PROXY_EXTRA_JWT_ISSUERS=https://oauth.mocklab.io/=https://oauth.mocklab.io/ \
           -e OAUTH2_PROXY_ALLOWED_GROUPS=fake \
           --network="host" \
           de09d99144e6
````

**Comment by Bananenbieger1234** [comment]
Hey,
we are facing the same problem... When is the v8 release planned ?
Are there any workarounds to this issue?

Thx.

**Comment by theGOTOguy** [comment]
@Bananenbieger1234 We created [our own fork](https://github.com/duke-crucible/oauth2-proxy) that resolves this issue for now.  If you want to see specifically what needs to happen and make your own fork, #1242 was my suggested short-term fix.

**Comment by NickMeves** [comment]
> Hey,
> we are facing the same problem... When is the v8 release planned ?
> Are there any workarounds to this issue?
> 
> Thx.

If your IdP allows it -- I recommend shifting the group based authorization to their layer (e.g. in OneLogin I would make a specific OIDC app that only allowed authentication from certain groups there).

Otherwise the fork PR is a good recommendation.

This won't likely have a holistic fix in this codebase for a while. It requires 2 major projects: plugable bearer providers in future structured configs + centralized authorization (e.g. Oso or OPA authz rules engine that is provider agnostic).

Even ☝️  will be challenging until we add ways to identify sources of a session -- otherwise it will be hard to make authz rules that are aware that some sessions that don't have groups (machine bearer sessions) are somehow more privileged and can skip group restrictions in other parts of your application (I guess you would have to make rules specifically tied to the subject of your machine users 🤔 )

**Comment by Bananenbieger1234** [comment]
> @Bananenbieger1234 We created [our own fork](https://github.com/duke-crucible/oauth2-proxy) that resolves this issue for now. If you want to see specifically what needs to happen and make your own fork, #1242 was my suggested short-term fix.

what a lifesaver 👍 @theGOTOguy . We did a fork ourselves and integrated your changes manually.

@NickMeves I would appreciate to get this PR merged. I don't see any reason not to merge this and we don't want to rely on our fork forever^^. This change seems to me quite compatible, because if groups are available you always want to have them in your session , right?
Please correct me if I am wrong, maybe I don't get it ..

**Comment by theGOTOguy** [comment]
I do think that we should either accept #1242 (or something like it, e.g. #1413) to allow group claims to be loaded for OIDC --extra-jwt-issuers, or own the fact that we're not going to until v8 by accepting this PR and explicitly disallowing it.  Whatever we do, we shouldn't let the status quo be "allow users to select options that we know cannot possibly work together."

**Comment by theGOTOguy** [comment]
Closing since the changed file has now been refactored such that my PR now has nontrivial merge conflicts.  I am not sure if `--allowed-group` still causes `--extra-jwt-issuers` to be invalid because the groups claim isn't loaded.

**Pull request closed without merging**